File: csharp/Workers/PaymentFailureWorker.cs
using Camunda.Orchestration.Sdk;

namespace Camunda.Training.CSharp.Workers;

public class PaymentFailureWorker(CamundaClient client) : Worker("payment-failure", client)
{
    public override async Task<object?> Handler(ActivatedJob job, CancellationToken ct)
    {
        Console.WriteLine("Handling payment failure");

        var variables = job.GetVariables<Dictionary<string, object>>();
        if (variables != null
            && variables.TryGetValue("orderId", out object? orderIdObj)
            && orderIdObj?.ToString() is string orderId)
        {
            // Notify the waiting Order Process that payment has failed.
            // No need to send variables back; only the orderId correlation is needed.
            await client.PublishMessageAsync(new MessagePublicationRequest
            {
                Name = "paymentFailedMessage",
                CorrelationKey = orderId,
                TimeToLive = (long)TimeSpan.FromMinutes(5).TotalMilliseconds,
            }, ct);

            Console.WriteLine("Message paymentFailedMessage sent");
        }
        else
        {
            Console.WriteLine("orderId not found or not a string in the variables");
        }

        // Returning null auto-completes the job with no variables.
        return null;
    }
}

File: java/src/main/java/io/camunda/training/workers/PaymentFailureWorker.java
package io.camunda.training.workers;

import io.camunda.client.CamundaClient;
import io.camunda.client.api.response.ActivatedJob;
import io.camunda.client.api.worker.JobClient;
import io.camunda.client.api.worker.JobHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class PaymentFailureWorker implements JobHandler {

    private static final Logger logger = LoggerFactory.getLogger(PaymentFailureWorker.class);

    private final CamundaClient camundaClient;

    public PaymentFailureWorker(CamundaClient camundaClient) {
        this.camundaClient = camundaClient;
    }

    @Override
    public void handle(JobClient client, ActivatedJob job) throws Exception {
        logger.info("Job handled: {}", job.getType());

        String orderId = (String) job.getVariablesAsMap().get("orderId");

        // Tell Order Process to continue
        camundaClient.newPublishMessageCommand()
                .messageName("paymentFailedMessage")
                .correlationKey(orderId)
                .send().join();

        // Complete the job in the Payment Process
        client.newCompleteCommand(job).send();
    }
}
